File: python/plugins/page_links_plugin.py
from pathlib import Path
from typing import List, Dict, Union

from models.document import Document
from plugins.md2html_plugin import Md2HtmlPlugin
from utils import relativize_relative_resource

MODULE_DIR = Path(__file__).resolve().parent


class PageLinksPlugin(Md2HtmlPlugin):
    def __init__(self):
        super().__init__()
        self.markers: List[str] = []
        self.page_links_handlers: List[PageLinkMetadataHandler] = []

    def accept_data(self, data):
        self.assure_accept_data_once()
        self.validate_data_with_file(data, MODULE_DIR.joinpath('page_links_schema.json'))
        markers = data.get("markers")
        self.markers = markers if markers else ["page"]

    def accept_document_list(self, docs: List[Document]):
        if not self.markers:
            return
        documents = {d.code: d.output_file for d in docs if d.code}
        if documents:
            pmh = PageLinkMetadataHandler(documents)
            self.page_links_handlers = [(pmh, m, False) for m in self.markers]

    def is_blank(self) -> bool:
        return not bool(self.page_links_handlers)

    def page_metadata_handlers(self):
        return self.page_links_handlers


class PageLinkMetadataHandler:
    def __init__(self, pages: Dict[str, str]):
        self.pages = pages

    def accept_page_metadata(self, doc: Document, marker: str, metadata_str: str, metadata_section,
                             visited_markers: Union[Dict[str, None]] = None):

        metadata_str = metadata_str.strip()
        destination_page_output = self.pages.get(metadata_str)
        if destination_page_output is None:
            return metadata_section
        else:
            return relativize_relative_resource(destination_page_output, doc.output_file)

File: java/src/main/java/world/md2html/plugins/PageLinksPlugin.java
package world.md2html.plugins;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectReader;
import com.fasterxml.jackson.databind.node.NullNode;
import world.md2html.options.argfile.ArgFileParseException;
import world.md2html.options.model.Document;
import world.md2html.utils.CheckedIllegalArgumentException;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

import static world.md2html.utils.JsonUtils.OBJECT_MAPPER;
import static world.md2html.utils.Utils.relativizeRelativeResource;

public class PageLinksPlugin extends AbstractMd2HtmlPlugin implements PageMetadataHandler {

    private List<String> markers;
    private List<PageMetadataHandlerInfo> pageLinksHandlers = new ArrayList<>();
    private Map<String, String> pages;

    @Override
    public void acceptData(JsonNode data) throws ArgFileParseException {
        assureAcceptDataOnce();
        validateInputDataAgainstSchemaFromResource(data, "plugins/page_links_schema.json");
        JsonNode markersNode = data.get("markers");
        if (markersNode == null) {
            this.markers = Collections.singletonList("page");
        } else if (markersNode instanceof NullNode) {
            this.markers = Collections.emptyList();
        } else {
            ObjectReader reader = OBJECT_MAPPER.readerFor(new TypeReference<List<String>>() {});
            ArrayList<String> list;
            try {
                list = reader.readValue(markersNode);
            } catch (IOException e) {
                throw new ArgFileParseException("Error reading plugin '" +
                        this.getClass().getSimpleName() + "' data: " + data);
            }
            if (list.isEmpty()) {
                this.markers = Collections.singletonList("page");
            } else {
                this.markers = list;
            }
        }
    }

    @Override
    public boolean isBlank() {
        return this.pageLinksHandlers.isEmpty();
    }

    @Override
    public List<PageMetadataHandlerInfo> pageMetadataHandlers() {
        return this.pageLinksHandlers;
    }

    @Override
    public String acceptPageMetadata(Document document, String marker, String metadata,
            String metadataSection) throws PageMetadataException {

        String destinationPageOutput = this.pages.get(metadata.trim());
        if (destinationPageOutput == null) {
            return metadataSection;
        } else {
            try {
                return relativizeRelativeResource(destinationPageOutput, document.getOutput());
            } catch (CheckedIllegalArgumentException e) {
                throw new PageMetadataException("Plugin '" + this.getClass().getSimpleName() +
                        "': Cannot relativize '" + destinationPageOutput + "' against '" +
                        document.getOutput() + "': " + e.getMessage());
            }
        }
    }

    @Override
    public void acceptDocumentList(List<Document> documents) {
        if (this.markers.isEmpty()) {
            return;
        }
        this.pages = documents.stream().filter(d -> !Objects.isNull(d.getCode()))
                .collect(Collectors.toMap(Document::getCode, Document::getOutput));
        if (!this.pages.isEmpty()) {
            this.pageLinksHandlers = this.markers.stream()
                    .map(m -> new PageMetadataHandlerInfo(this, m, false))
                    .collect(Collectors.toList());
        }
    }

}
